### PR TITLE
Add page translation capability

### PR DESCRIPTION
**Pull request by sabertooth-cat**
Note: apertium-html-tools have also been modified.

https://codein.withgoogle.com/dashboard/task-instances/5480355175858176/?sp-page=1

### CODE REVIEW

File: translation.py [hunk 1]
```diff
@@ -287,3 +287,17 @@ def translateDoc(fileToTranslate, fmt, modeFile):
     mode = os.path.splitext(os.path.basename(modeFile))[0]
     return Popen(['apertium', '-f', fmt, '-d', modesdir, mode],
                  stdin=fileToTranslate, stdout=PIPE).communicate()[0]
+
+def translatePage(url, modeFile):
+    modesdir = os.path.dirname(os.path.dirname(modeFile))
+    mode = os.path.splitext(os.path.basename(modeFile))[0]
+    response = urllib.request.urlopen(url)
+    text = response.read().decode('utf-8')
```
Comment (unhammer) [inline translation.py]: needs the right encoding depending on what the *page* says it should be

File: translation.py [hunk 2]
```diff
@@ -287,3 +287,17 @@ def translateDoc(fileToTranslate, fmt, modeFile):
...
+    
+    
+    text = text.replace('href="/',  'href="{uri.scheme}://{uri.netloc}/'.format(uri=urlparse(url)))
+    text = re.sub(r'a([^>]+)href=[\'"]?([^\'" >]+)', 'a \\1 href="#" onclick=\'window.parent.translateLink("\\2");\'', text)
+    logging.info(text)
+    return Popen(['apertium', '-f', 'html', '-d', modesdir, mode],
+                 stdin=PIPE, stdout=PIPE).communicate(text.encode('utf-8'))[0]
```
Comment (unhammer) [inline translation.py]: This'll block for the duration of the translation, please use the same system as the regular "translate" function (in fact, you might be able to just call that)

**Comment by unhammer** [comment]
please merge in master as well so we can see test results

**Comment by sabertooth-cat** [comment]
How do I merge in master?

**Comment by unhammer** [comment]
if you haven't already added goavki as a remote,

    git remote add goavki https://github.com/goavki/apertium-apy

then

    git fetch remote
    git merge goavki/master

File: servlet.py [hunk 3]
```diff
@@ -342,6 +350,45 @@ def get(self):
...
+                encoding = chardet.detect(data)["encoding"]
+            else:
+                encoding = 'utf-8'
+            text = data.decode(encoding)   
+            text = text.replace('href="/',  'href="{uri.scheme}://{uri.netloc}/'.format(uri=urlparse(url)))
+            text = re.sub(r'a([^>]+)href=[\'"]?([^\'" >]+)', 'a \\1 href="#" onclick=\'window.parent.translateLink("\\2");\'', text)
+            logging.info(text)
+            commands = [['apertium', '-d', modesdir, '-f', "html", mode]]
+            logging.info("1")
+            result = yield translation.translateSimple(text, commands)
```
Comment (unhammer) [inline servlet.py]: look at what `get` in TranslateHandler does – this `get` should do the same, except for turning the url into a text first. In particular the actual call to translate needs to use `translated = yield pipeline.translate(toTranslate)`. 

We want to have caps on how many pipelines of a pair can be open at once, so the server admin doesn't have to worry about massive amounts of translation requests creating a new pair on every request. Imagine 100 people at the same time translate web sites; we don't want to start 100 pipelines at once and eat all our RAM (some of those analysers are huge). The `TranslateHandler.get` method avoids this by keeping at most N pipelines open at any one time.

File: translation.py [hunk 4]
```diff
@@ -62,7 +62,8 @@ def __del__(self):
     def translate(self, toTranslate):
         with self.use():
             all_split = splitForTranslation(toTranslate, n_users=self.users)
-            parts = yield [translateNULFlush(part, self) for part in all_split]
+            # parts = yield [translateNULFlush(part, self) for part in all_split]
+            parts = yield [translateNULFlush(toTranslate, self)]
```
Comment (unhammer) [inline translation.py]: as I said on IRC, add an optional arg "nosplit=False" here; we only want this non-splitting hack if "nosplit=True"

**Pull request merged**